server/src/actors/project.rs

```diff
@@ -412,35 +411,125 @@ pub struct PublicKeyConfig {
 }
 
 /// Data for applying rate limits in Redis
+#[derive(Debug, Clone)]
+pub enum Quota {
+    RejectAll(RejectAllQuota),
+    Redis(RedisQuota),
+}
+
+mod __quota_serialization {
+    use super::{Quota, RedisQuota, RejectAllQuota};
+
+    use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
+
+    #[derive(Serialize, Deserialize, Default)]
+    #[serde(default)]
+    struct QuotaSerdeHelper {
+        #[serde(skip_serializing_if = "Option::is_none")]
+        limit: Option<u64>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        reason_code: Option<String>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        prefix: Option<String>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        subscope: Option<String>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        window: Option<u64>,
+    }
+
+    impl Serialize for Quota {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            match *self {
+                Quota::RejectAll(RejectAllQuota { ref reason_code }) => QuotaSerdeHelper {
+                    limit: Some(0),
+                    reason_code: reason_code.clone(),
+                    ..Default::default()
+                },
+                Quota::Redis(RedisQuota {
+                    limit,
+                    ref reason_code,
+                    ref prefix,
+                    ref subscope,
+                    window,
+                }) => QuotaSerdeHelper {
+                    limit,
+                    reason_code: reason_code.clone(),
+                    prefix: Some(prefix.clone()),
+                    subscope: subscope.clone(),
+                    window: Some(window),
+                },
+            }
+            .serialize(serializer)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Quota {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let helper = QuotaSerdeHelper::deserialize(deserializer)?;
+
+            match (
+                helper.limit,
+                helper.reason_code,
+                helper.prefix,
+                helper.subscope,
+                helper.window,
+            ) {
+                (Some(0), reason_code, None, None, None) => {
+                    Ok(Quota::RejectAll(RejectAllQuota { reason_code }))
+                }
+                (limit, reason_code, Some(prefix), subscope, Some(window)) => {
+                    Ok(Quota::Redis(RedisQuota {
+                        limit,
+                        reason_code,
+                        prefix,
+                        subscope,
+                        window,
+                    }))
+                }
+                _ => Err(D::Error::custom(
+                    "Could not deserialize, no variant matched",
```

Comment:
_nit_: Please write something about `Quota` in this message. Unfortunately, Serde otherwise gives no indication of why deserialization fails and this would make it quite hard to debug.